### PR TITLE
fix(Core/Spells): Master of Elements

### LINKED ISSUE #6035: [Mage Talents] Master of Elements not restoring mana on critical strike
#### WHAT CLIENT DO YOU PLAY ON?

- [ ] enGB
- [x] enUS
- [ ] other (specify)

##### FACTION

- [ ] Alliance
- [x] Horde

##### CONTENT PHASE:

- [x] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

Original report: https://github.com/chromiecraft/chromiecraft/issues/694

##### CURRENT BEHAVIOUR:

Succesful spell criticals do not always trigger Master of Elements (https://wowgaming.altervista.org/aowow/?spell=29076).
The following spells never trigger it:
Arcane Explosion - https://wowgaming.altervista.org/aowow/?spell=8439
Arcane Missiles -  https://wowgaming.altervista.org/aowow/?spell=8416
Blizzard - https://wowgaming.altervista.org/aowow/?spell=8427

And for other spells it doesn't always trigger:
https://i.imgur.com/cXyhNxH.png
https://i.imgur.com/l9yQJl0.png

##### EXPECTED BLIZZLIKE BEHAVIOUR:

Every spell should be able to benefit from Master of Elements as of patch 3.0.2: https://wowwiki-archive.fandom.com/wiki/Patch_3.0.2

##### STEPS TO REPRODUCE THE PROBLEM:

1. Cast a spell
2. Land a critical
3. ???
4. No mana

##### EXTRA NOTES:

Master of Elements: https://wowgaming.altervista.org/aowow/?spell=29076

##### AC HASH/COMMIT:

https://github.com/chromiecraft/azerothcore-wotlk/commit/e50754d804987ec93792b6a4d113913ab018189b

##### OPERATING SYSTEM:

Ubuntu 20.04

##### MODULES:

- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-LevelUpReward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-CarbonCopy](https://github.com/55Honey/Acore_CarbonCopy)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)

##### OTHER CUSTOMIZATIONS:

None.

##### SERVER:

ChromieCraft

Confirmed and reproduced on a local AC server (AC HASH/COMMIT: ea764812d14a):

WoWScrnShot_052521_084239
WoWScrnShot_052521_084330

There are some little mana restoring called 'burnout' but i don't think it's related with the mentioned passive, since the % does not correspond with the amount of mana of the ability used.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/98658095-mage-talents-master-of-elements-not-restoring-mana-on-critical-strike?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UltraNix**
Proc now also from arcane spells.
Properly proc from periodic spells.
Fixed #6035.

## Changes Proposed:
-  Should proc off from arcane spells.
-  Spells like Arcane Missiles or Blizzard return mana per tick.

## Issues Addressed:

- Closes #6035
- Closes https://github.com/chromiecraft/chromiecraft/issues/694

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

1. Cast a spell
2. Land a critical
3. Observer your combat log.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by pklloveyou** [comment]
Failed the test and did not recover mana

File: src/server/game/Spells/SpellMgr.cpp [hunk 1]
```diff
@@ -5568,6 +5568,9 @@ void SpellMgr::LoadDbcDataCorrections()
         spellInfo->AttributesEx3 |= SPELL_ATTR3_SUPRESS_TARGET_PROCS;
     });
 
+    ApplySpellFix({ 29069  }, [](SpellEntry* spellInfo) { spellInfo->EffectBasePoints[0] = 5000;
+    });
```
Comment (FrancescoBorzi) [inline src/server/game/Spells/SpellMgr.cpp:5572]: is this weird indentation intended?

Comment (UltraNix) [reply src/server/game/Spells/SpellMgr.cpp:5572]: It should not be there,

**Comment by UltraNix** [comment]
@pklloveyou Are you sure you had the talent active and your hits were critical?

**Comment by pklloveyou** [comment]
image
Then you don't restore mana after critical hit with an arcane missile

**Comment by pklloveyou** [comment]
Frost spell triggers the elemental master and restores mana

**Comment by pklloveyou** [comment]
Frost spell triggers the elemental master and restores mana
Frost spell before PR cannot restore mana

**Comment by pklloveyou** [comment]
“snowstorm”  cannot restore mana before PR

**Comment by pklloveyou** [comment]
AND Too little mana restored after pr

**Comment by UltraNix** [comment]
What do you mean by "Too little mana restored"?  Speciific spell?
Remember that it's calculated on based mana cost (without any modifiers).

**Comment by UltraNix** [comment]
And whats is `snowstorm` spell?

**Comment by pklloveyou** [comment]
snowstorm speel=42940

**Comment by pklloveyou** [comment]
Before pr
image
image
620*0.3=186

**Comment by pklloveyou** [comment]
Wait a minute. I'm building the code

**Comment by UltraNix** [comment]
https://wowpedia.fandom.com/wiki/Master_of_Elements
```
For channeled spells that deal damage in ticks such as  [Arcane Missiles] and  [Blizzard], the mana returned is based on the tick that critically hit, and not the entire spell. For example, Arcane Missiles in its higher levels deals damage in 5 ticks: it will only return 15/30% of 1/5th of the cost of Arcane Missiles and not its total.
```
620/8 ticks * 30% of restored mana

**Comment by pklloveyou** [comment]
Before pr
image
image

After pr
image
image

**Comment by pklloveyou** [comment]
image

image

**Comment by pklloveyou** [comment]
image
image

**Comment by UltraNix** [comment]
So it works okay.

**Comment by pklloveyou** [comment]
Arcane spells do not restore mana at all

**Comment by pklloveyou** [comment]
The above tests are all instant spells,

**Comment by UltraNix** [comment]
Arcane spells should be okay now.

**Comment by pklloveyou** [comment]
Let me try

**Comment by pklloveyou** [comment]
The restored mana is still wrong,
spell 42921       718*0.3=215.4
image

**Comment by pklloveyou** [comment]
Oh, I see. It's back to normal after resetting the talent

**Comment by pklloveyou** [comment]
Tested

**Comment by pklloveyou** [comment]
Thank you very much for your patient answers and contributions

**Comment by pklloveyou** [comment]
@FrancescoBorzi

**Comment by robbie018** [comment]
<h1>Before fix</h1>
Some spells didn't gave the proc, I've tested it a few times, but after a few crits, still no proc. And yes I had the talent selected.

https://user-images.githubusercontent.com/65018032/127027917-5d275ee0-3393-4c75-8f96-df2a6d7e210b.mp4

<h1>After fix</h1>
It works. So after each crit spell I indeed get the mana back.

https://user-images.githubusercontent.com/65018032/127028103-00c13863-8098-455a-8055-fbe6689180bc.mp4

**Review by Helias — APPROVED** [review]

**Pull request merged**